Segment: add a delay between 3 Track events after Group

0x656b694d opened the pull request:
## Description

As Segment doesn't guarantee the processing order of the events, and we need to send Group+Track in order, we send 3 consecutive Track events with 2s delay between to raise the chances for the group properties to be captured.

This code is used today only in Fleet Manager when assigning Tenant properties to a new Tenant group.

This change makes sense only if max batch size is set to 1, so that all events are sent immediately and separately.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

Unit tests.

0x656b694d commented:
Current dependencies on/for this PR:
* master
  * **PR #5195** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/5195" target="_blank"><img src="https://static.graphite.dev/graphite-32x32.png" alt="Graphite" width="10px" height="10px"/></a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/5195?utm_source=stack-comment).

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -209,12 +211,27 @@ func (t *segmentTelemeter) Group(props map[string]any, opts ...telemeter.Option)
+			Event:       "Group Properties Updated",
+			Context:     group.Context,
 		}
+		go func() {
```
Instead of adding another dependency, we could do the following instead:
- use a `*timer.Ticker` which will have an interval of two seconds.
- within the spawned go routine, we will have essentially something like this:
```go
var trackCalls int
for trackCalls < 3 {
		select {
		case <-t.ticker.C:
			if err := t.client.Enqueue(track); err != nil {
				log.Error("Cannot enqueue Segment track event: ", err)
			}
                // Theoretically, if we had to respect context cancellation, we would do so here.
		}
}
```
Within the initialization of the tracker we would start the ticker, and within the `Stop()` implementation of the `telemeter.Telemeter` interface we would stop this. I personally don't think it's worth to add a dependency for this.

0x656b694d replied:
True, the same could be done with a timer. Please have a look at the updated code.
I found a use of `github.com/benbjohnson/clock` in operator tools, so thought it's fine to use it here.

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -209,12 +212,31 @@ func (t *segmentTelemeter) Group(props map[string]any, opts ...telemeter.Option)
+			Context:     group.Context,
 		}
+		go func() {
+			if ti == nil {
+				ti = time.NewTicker(2 * time.Second)
+			}
```
We should probably do this either a) behind a lock or b) outside of the go routine. Currently, this would create a race condition between two go routines accessing this value.

0x656b694d replied:
There wouldn't be a race for this variable, as it's bound to a call.
There is a race though with Segment internal channel, if we close the client while the goroutine is still running.

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -209,12 +212,31 @@ func (t *segmentTelemeter) Group(props map[string]any, opts ...telemeter.Option)
+					log.Error("Cannot enqueue Segment track event: ", err)
+				}
+			}
+			ti.Stop()
```
Hm, we could use a shared ticker throughout the telemeter by having it as a property on the telemeter struct and initialize it in new / start.

If you want to do one on a per-track, maybe factor this inline goroutine into a separate function that takes in a ticker that will be created within the current track function?

0x656b694d replied:
Please take a look at the per-track timer.

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -18,6 +19,8 @@ type segmentTelemeter struct {
 	clientID   string
 	clientType string
+
+	stopMux *sync.Mutex
```
Maybe a `sync.RWMutex` is more fitting?

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -155,7 +165,7 @@ func (t *segmentTelemeter) makeContext(o *telemeter.CallOptions) *segment.Contex
 
 func (t *segmentTelemeter) Identify(props map[string]any, opts ...telemeter.Option) {
-	if t == nil {
+	if t == nil || t.client == nil {
```
Shouldn't this be also guarded behind a mutex then? Theoretically, the same datarace could happen here as well?

0x656b694d replied:
Yeah, a different race, but true.

dhaus67 commented on pkg/telemetry/phonehome/segment/segment.go:
```diff
@@ -241,7 +256,7 @@ func (t *segmentTelemeter) group(props map[string]any, ti *time.Ticker, opts ...
 
 func (t *segmentTelemeter) Track(event string, props map[string]any, opts ...telemeter.Option) {
-	if t == nil {
+	if t == nil || t.client == nil {
```
Same as above w.r.t. doing this guarded behind a mutex.

dhaus67 reviewed (approved):
Just minor comments regarding moving to `RWMutex` and other places where we should use a mutex.

0x656b694d commented:
@dhaus67 , behold, I removed the mutex. Segment SDK doesn't need to be synchronized, so I refactored the code so that the test execute the synchronous pieces.

dhaus67 reviewed (approved):
Sorry for the late review, LGTM 🚢
